charms/worker/k8s/src/charm.py

```diff
@@ -197,6 +196,32 @@ def _apply_proxy_environment(self):
             with open("/etc/environment", mode="w", encoding="utf-8") as file:
                 file.write("\n".join([f"{k}={v}" for k, v in current_env.items()]))
 
+    def get_cluster_name(self) -> str:
+        """Craft a unique name for the cluster once joined or bootstrapped.
+
+        Note: It won't change for the lifetime of the unit.
+
+        Returns:
+            the cluster name.
+        """
+        unit, node = self.unit.name, self.get_node_name()
+        if not self._stored.cluster_name:
+            if self.lead_control_plane and self.api_manager.is_cluster_bootstrapped():
+                # TODO: replace with API call once available from the snap
+                self._stored.cluster_name = str(uuid.uuid4())
+            elif not (relation := self.model.get_relation("cluster")):
+                pass
+            elif any(
+                [
+                    self.is_control_plane and self.api_manager.is_cluster_bootstrapped(),
+                    self._is_node_present(),
+                ]
+            ):
```

Comment:
If i'm a control-plane unit that's bootstrapped, or my node is present in the kubernetes cluster, then i'm clustered finally and should set my `cluster_name` based on what's received from the leader control-plane unit who gave me the join token